### ISSUE TITLE
@ is replaced with %40 in Username when using E-Mail for Login

### ISSUE THREAD

**Issue by BornToBeRoot** [privacyidea/keycloak-provider#67]
Hi,

when using the e-mail attribut as login in keycloak, the username is passed with `%40` to the ldap server. Because of this encoding, the user is not found in the ldap resolver.

If the uid doesn't contain an `@` the, the login works.

Greetings

**Comment by BornToBeRoot** [comment]
```
11:45:39,239 INFO  [org.privacyidea.authenticator.PrivacyIDEAAuthenticator] (default task-2) test.benutzer@domain.de
11:45:39,239 INFO  [org.privacyidea.authenticator.PrivacyIDEAAuthenticator] (default task-2) org.keycloak.authentication.AuthenticationProcessor$Result@63d0c4d2

```

```
[2021-06-28 13:45:39,134][68006][139867774859008][DEBUG][privacyidea.lib.resolvers.LDAPIdResolver:1314] Doing start_tls
[2021-06-28 13:45:39,169][68006][139867774859008][DEBUG][privacyidea.lib.resolvers.LDAPIdResolver:654] login name filter: 'uid=test.benutzer%40domain.de'
[2021-06-28 13:45:39,170][68006][139867774859008][DEBUG][privacyidea.lib.resolvers.LDAPIdResolver:662] Searching user 'test.benutzer%40domain.de' in LDAP.
[2021-06-28 13:45:39,173][68006][139867774859008][DEBUG][privacyidea.lib.user:244] user 'test.benutzer%40domain.de' not found in resolver 'ldap.domain.de'
```

**Comment by nilsbehlen** [comment]
Hi,
all parameters are url encoded, because otherwise the input could be tampered with, e.g. adding '&serial=...' to the password.
I will check with the server side on how to handle this.

**Comment by SorinMiron** [comment]
https://community.privacyidea.org/t/login-from-keycloak-to-privacyidea-with-email/2013/6

Yaaaayy, i'm not crazy!
I'm struggling with the same issue.

**Comment by nilsbehlen** [comment]
https://github.com/privacyidea/privacyidea/issues/2800#issuecomment-873858994

**Comment by Soapfedan** [comment]
I'm facing the same issue. Keycloak send request to Privacyidea with %40 instead of @. So privacyidea returns to me the **ERR904: The user can not be found in any resolver**. Are there any news?

**Comment by nilsbehlen** [comment]
It will be fixed in the next version of privacyIDEA, see the link to the issue in my previous comment.
If you want it fixed now, edit this line https://github.com/privacyidea/sdk-java/blob/c669f1b77144427026e0f7f7a812f9c9ab6928d1/src/main/java/org/privacyidea/Endpoint.java#L137 and rebuild the jar.

**Comment by Soapfedan** [comment]
Ok so the 0.6.1 deployment jar doesn't contain the new line. Is it correct?
So tell me if this procedure is correct:
1) download the source code
2) edit the sdk wrong line
3) build the new jar

Is it ok?

**Comment by nilsbehlen** [comment]
No, 0.6.1 does not contain it. Note that editing the code of the provider is not how things are intended to be, so it is only a **TEMPORARY** workaround if you need it urgently. The code that has to be changed is located in the java-sdk which is a submodule of this repo.

1. Clone the respository and pull the submodules.
2. Change the code in the sdk. (just comment lines 135-141)
3. Rebuild the sdk (mvn clean install in \lib\sdk-java\)
4. Rebuild the provider
5. The jar is in \target

**Comment by Soapfedan** [comment]
Thanks it works!!

**Comment by cbr46** [comment]
Hi @nilsbehlen ,
same problem here. I've installed privacyIDEA 3.6.1 from the Ubuntu repo and fetched the latest KeyCloak Provider (v. 0.6.1) and still get the error messages. I can't follow your steps, do I have to manually compile privacyIDEA or the KeyCloak Provider?
I'm running on KeyCloak 13.0.1.

**Comment by nilsbehlen** [comment]
@cbr46 you will have to edit the code of the keycloak provider (this repo) and built it again, as i described in the previous post.

**Comment by cbr46** [comment]
Hi @nilsbehlen,

is there a new release planned after this issue was closed?

**Comment by nilsbehlen** [comment]
Hey, the issue will be fixed in the next server version (v3.7) https://github.com/privacyidea/privacyidea/issues/2800
